### DISCUSSION TITLE
Auth error: {401: invalid claim: missing sub}

### DISCUSSION THREAD

**Discussion by monicakh** [supabase/supabase#21347] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/auth-error-401-invalid-claim-missing-sub--AFwMR)._

The missing sub claim error is returned when `supabase.auth.getUser()` is called with an invalid JWT in the session or when the user attempts to register/sign in but hasn't completed the sign in when the `getUser` call is made.

A common pitfall, is inadvertently using a Supabase API key (such as the publishable or secret keys) instead of the Supabase Auth access token.

**Why Does This Happen?**

*   The Supabase API keys are designed for different purposes and, while they are recognized by the Supabase Auth system, they do not carry the sub claim. The sub claim is essential as it encodes the user ID, which is a mandatory field for authentication processes. This mistake leads to the "missing sub claim" error because the system expects a token that contains user identification information.

**How to Avoid This Issue:**

*   Ensure that the token being passed to `supabase.auth.getUser()` is, indeed, an Auth access token and not one of the API keys.

*   Are you creating the client on a per-request basis or are you creating a global client to be shared? If you're creating the client on a per-request basis, then you need to pass the session with the user's JWT from the client to the server somehow. This can be done by sending the user's JWT in a header like an `Authorization: Bearer <user_jwt>` . You can then get this header and call `supabase.auth.getUser(user_jwt)` with the user's JWT.

*   Examine how the Supabase client is being initialized, especially in server-side scenarios.

**Comment by layerzzzio** [comment]
Hi @monicakh 

I am having this issue with NextJS when executing this server action by clicking on a like button:

```
export async function optimisticLikeReaction(
  aiPhotoName: string,
  isLike: boolean
) {
  const supabase = createClient();
  const user = await getUser();
  const { error } = await supabase.from("photo_likes").upsert({
    ai_picture_name: aiPhotoName,
    user_id: user.id,
    is_like: isLike,
  });

  if (error) {
    console.log(`error=${error}`);
    return error;
  }

  // Optionally handle the success case
  // return data;
}
```

The function await getUser() in optimisticLikeReaction, designed as follows, is failing :

```
export async function getUser(): Promise<User> {
  const supabase = createClient();

  try {
    const { data, error } = await supabase.auth.getUser();

    if (error) {
      throw error;
    }

    if (!data || !data.user) {
      throw new Error("User not found");
    }

    // Assuming data.user is of type UserResponse
    // Otherwise, map the properties accordingly
    return data.user;
  } catch (error) {
    // Handle or rethrow the error
    console.error("Error fetching user info:", error);
    throw error;
  }
}
```

The supabase variable in function `getUser` is created from createClient (server) from supabase/ssr package:

```
import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { cookies } from "next/headers";

export function createClient() {
  const cookieStore = cookies();

  return createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options });
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: "", ...options });
          } catch (error) {
            // The `delete` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  );
}
```

I also have a middleware.ts that executes this:
```
import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { NextResponse, type NextRequest } from "next/server";

export async function updateSession(request: NextRequest) {
  let response = NextResponse.next({
    request: {
      headers: request.headers,
    },
  });

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return request.cookies.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value,
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value,
            ...options,
          });
        },
        remove(name: string, options: CookieOptions) {
          request.cookies.set({
            name,
            value: "",
            ...options,
          });
          response = NextResponse.next({
            request: {
              headers: request.headers,
            },
          });
          response.cookies.set({
            name,
            value: "",
            ...options,
          });
        },
      },
    }
  );

  await supabase.auth.getUser();

  return response;
}
```

Do you see anything wrong with my process?

Note: I followed the SSR guide here https://supabase.com/docs/guides/auth/server-side/nextjs.

Thank you so much for your help
Best

**Comment by Flohhhhh** [comment]
This error occurs when trying to sign up on the NextJS Supabase Starter here: https://github.com/vercel/next.js/commits/canary/examples/with-supabase without changing anything. It works once in a while, then fails, presumably because of the rate limit.

The result is some sort of looping issue like this:

![image](https://github.com/supabase/supabase/assets/48927090/bf147e5c-d452-4b54-824d-8dead1d61733)

**Comment by PiessensLander** [comment]
I'm getting this issue on my NextJS App Router project as well. The first login usually works as expected: a cookie gets set, the user is redirected to a dashboard and can see the relevant data. However, after logging out and trying to log in again, the server responds with the `401: missing sub` error. The logout function seems to work fine since the cookie gets deleted. I'm not sure what's going wrong but it's a real dealbreaker because now my application is useless.

**Comment by hf** [comment]
This comes from the fact that the anon Supabase API key is somehow "leaking" through to the `GET /user` endpoint.

**Comment by curiosbasant** [comment]
I'm really stuck at this issue. Have spent my whole day and couldn't found any solution for it. 😔
